sklearn/cluster/_birch.py

```diff
@@ -698,9 +696,6 @@ def _global_clustering(self, X=None):
                 "n_clusters should be an instance of ClusterMixin or an int"
             )
 
-        # To use in predict to avoid recalculation.
-        self._subcluster_norms = row_norms(self.subcluster_centers_, squared=True)
-
```

Comment:
The fact that we no longer reuse the precomputed row norms does not cause a performance problem? Or is the new code so much faster that this is no longer necessary? 


Wouldn't it make sence to make it possible to still pass precomputed `Y_norm_squared` to `pairwise_distances_argmin` with `metric="fast_sqeuclidean"`?

Reply:
> The fact that we no longer reuse the precomputed row norms does not cause a performance problem? Or is the new code so much faster that this is no longer necessary?

I haven't tested it yet, this patch came mainly as an adaptation for `pairwise_distances_argmin`.

> Wouldn't it make sence to make it possible to still pass precomputed Y_norm_squared to pairwise_distances_argmin with metric="fast_sqeuclidean"?

It's doable, as of now they get computed in an sub-optimal way [here](https://github.com/scikit-learn/scikit-learn/blob/15c4150c911f463a9eaacf11e1b815f1d16345b9/sklearn/metrics/_parallel_reductions.pyx#L730).

As a side note, this call in Birch is [the only case of passing `Y_norm_squared` to `pairwise_distances_argmin` in the code-base](https://github.com/scikit-learn/scikit-learn/search?q=Y_norm_squared).

Reply:
For the sake of precision: `row_norms` performs under the hood the same operation as the ones used in this new implementation there using `einsum` (singly-threaded but memory-efficient), namely:

```python
einsum('ij,ij->i', Y, Y)
```

Benchmarks show that the proposed implementation for `ArgKmin` makes Birch faster, see my comment: https://github.com/scikit-learn/scikit-learn/pull/20254#issuecomment-882383882.